editorModule.js

```diff
@@ -31,30 +32,272 @@ export default class EditorModule{
+                triggerCharacters: ['.', ' '],
+
+                // Function to generate autocompletion results
+                provideCompletionItems: function(model, position, token) {
+
+                    // Return if this is not the javascript editor
+                    if(model.id != '$model3') {
+                        return
+                    }
 
-        //27, javascript
-        const javascriptLoader = await monaco.languages.getLanguages()[27].loader();
+                    // Get current word being typed
+                    var active_typing = getActiveWord(model, position);
 
-        monaco.languages.setMonarchTokensProvider('FormItWSM', javascriptLoader.language);
+                    // If the last character typed is a period then we need to look at member objects of obj
+                    var is_member = active_typing.indexOf('.') >= 0
 
-        try {
-            const autocompleteURL = "https://formit3d.github.io/FormItExamplePlugins/docs/autocomplete.json",
-                response = await fetch(autocompleteURL),
-                autocompleteJSON = await response.json()
+                    // Array of autocompletion results
+                    var result = [];
+
+                    // Existing value
+                    var curValue = model.getValue();
+
+                    // Used for generic handling between member and non-member objects
+                    var last_token = obj; 
+                    var prefix = ''; 
+
+                    // Get a list of all members, and the prefix
+                    var parents = active_typing.split("."),
+                        last_word = parents[parents.length - 1]
+                        parents = parents.splice(0, parents.length - 1);
+
+                    if(parents.length) {
+                        last_token = obj[parents[0]];
+                        prefix = parents[0]; 
+                    }
+
+                    // Loop through all the parents the current one will have (to generate prefix)
+                    for (var i = 1; i < parents.length; i++) { 
+                        if (last_token.hasOwnProperty(parents[i])) { 
+                            prefix += '.' + parents[i]; 
+                            last_token = last_token[parents[i]];
+                        } else { 
+                            // Not valid
+                            return result;
+                        }
+                    }
+
+                    // Append a . if we are inside the object
+                    if(prefix) {
+                        prefix += '.';
+                    }
+
+                    // Return no results if object reference is undefined
+                    if(last_token === undefined) {
+                        return;
+                    }
+
+                    // Get all the keys of the object
+                    var all_token_names = Object.getPropertyNames(last_token);
```

Comment:
Think you want to use Object.keys() https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys

You're just trying to get the props of an object? If so, then use that. Then you don't need the prototype checking logic below.

Reply:
When I first implemented this, I tried Object.keys but it didn't return all possible property names for the object. I did some searching and found the library function "getPropertyNames" I added in editor.html. I can revisit this to see if I can reduce it further, or remove it entirely